src/adapter/src/coord/timestamp_selection.rs

```diff
@@ -356,6 +369,34 @@ pub trait TimestampProvider {
             candidate.join_assign(&real_time_recency_ts);
         }
 
+        let mut session_oracle_read_ts = None;
+        if isolation_level == &IsolationLevel::StrongSessionSerializable {
+            if let Some(timeline) = &timeline {
+                if let Some(oracle) = session.get_timestamp_oracle(&timeline) {
+                    let session_ts = oracle.read_ts();
+                    candidate.join_assign(&session_ts);
+                    session_oracle_read_ts = Some(session_ts);
+                }
+            }
+
+            // When advancing the read timestamp under Strong Session Serializable, there is a
+            // trade-off to make between freshness and latency. We can choose a timestamp close the
+            // `upper`, but then later queries might block if the `upper` is too far into the
+            // future. We can chose a timestamp close to the current time, but then we may not be
+            // getting results that are as fresh as possible. As a heuristic, we choose the minimum
+            // of now and the upper, where we use the global timestamp oracle read timestamp as a
+            // proxy for now. If upper > now, then we choose now and prevent blocking future
+            // queries. If upper < now, then we choose the upper and prevent blocking the current
+            // query.
+            if when.can_advance_to_upper() && when.can_advance_to_timeline_ts() {
+                let mut advance_to = largest_not_in_advance_of_upper;
+                if let Some(oracle_read_ts) = oracle_read_ts {
+                    advance_to = std::cmp::min(advance_to, oracle_read_ts);
+                }
+                candidate.join_assign(&advance_to);
```

Comment:
What happens if we don't advance at all and only use the session_ts above?

Reply:
There are essentially two things that move the session timestamp oracle's read timestamp forward:
1. Writes to tables.
2. This call to `join_assign`.

Without this, nothing else moves the timestamp forward in the absence of write and we continuously read from the same timestamp over and over.

Reply:
The advancing here is the big difference between Strict Serializable and Strong Session Serializable. Strict Serializable timestamps advance with real time while Strong Session Serializable timestamps advance based on what we've previously read in the same session.

Reply:
Gotcha, so this is the thing that optimizes for freshness without sacrificing availability.